### DISCUSSION TITLE
How to integrate react-draft-wysiwyg with react-hook-form

### DISCUSSION THREAD

**Discussion by mattbaratella** [react-hook-form/react-hook-form#1764] [Q&A] [answered]
Hi guys, 

I need some help. I was using react-final-form in my project but I've decided to change for react-hook-form. By the way, I love it, but I got stuck. :/

On my page, I'm using an editor to collect some info from a user. At the moment I'm using react-draft-wysiwyg.
Here the code:

**parentComponent.js**
```
/**
 * react-hook-form library
*/
const {
    register,
    handleSubmit,
    control
} = useForm({
    mode: 'onChange'
});

<form onSubmit={handleSubmit(onSubmit)}>
    <Controller
        as={<WYSIWYGEditor />}
        name='editor'
        control={control}
        onChange={([ event ]) => event.target.value}
    />
</form>
```

**WYSIWYGEditor.js**
```
const WYSIWYGEditor = ({ input }) => {
	const [ editorState, setEditorState ] = useState(EditorState.createEmpty());
	const onEditorStateChange = editorState => {
		setEditorState(editorState);
	};

	return (
		<React.Fragment>
			<div className={classes.root}>
				<Editor
					editorState={editorState}
					wrapperClassName='wrapper-class'
					editorClassName='editor-class'
					onEditorStateChange={onEditorStateChange}
				/>
			</div>
		</React.Fragment>
	);
};

export default WYSIWYGEditor;
```

**_PLEASE NOTE:_** 
The input prop comes from the coding with react-final-form.  The input was passing the characters I was typing. So, if I leave `input` as props it fails because it doesn't exist. React-hook-form doesn't pass an input. 
I've changed that with `props`:
```
const WYSIWYGEditor = props=> {
    console.log(props)
```

and I get the following in the console.log when I type anything:

```
{name: "editor", value: undefined, onChange: ƒ}
```

As you can see, value is `undefined`. How can I structure the `Controller` in order to pass a value each time I type something in the editor?

Thanks for your help
Matt

**Comment by bluebill1049** [comment, marked as answer]
provide a simple codesandbox.

**Reply by mattbaratella** [reply]
Hi Bill,

Below I provide the final-form version, which is returning the value of input when the user types, and the react-hook-form version where the problem is.

**React-Final-Form version**
Check the console log in the WYSIWYG.js file. You'll notice that the value is returned while typing.
https://codesandbox.io/s/flamboyant-frost-dspdl?file=/src/App.js

**React-Hook-Form version**
Same here, if you check the WYSIWYG.js file, value (in the console.log) is always undefined.
https://codesandbox.io/s/trusting-mountain-k24ys?file=/src/App.js

Let me know
Thanks

**Reply by bluebill1049** [reply]
https://codesandbox.io/s/wizardly-williams-2cjoe?file=/src/App.js

**Reply by bluebill1049** [reply]
i am using v6 render props.

**Reply by mattbaratella** [reply]
Hi Bill,

Thanks for that. 
But because I have to convert text into HTML to be saved in the DB, the proposed solution didn't fit the use of drafts-to-html.

I had a look again at my old code and compared the passed `input` of react-final-form (RFF), with the `props` of react-hook-form (RHF).

I didn't notice before that they are exactly the same. So, I decided to change the Controller to this: (removed the onChange)
```
<Controller
    as={<WYSIWYGEditor />}
    name='editor_content'
     control={control}
/>
```
and replaced in WYSIWYG.js file the line `return input.onChange(draftToHtml(convertToRaw....` with `return props.onChange(draftToHtml(convertToRaw....`.

Boom, everything worked fine. I should have done the comparison before.
Shame on me.

Thanks for your help.
Matt

**Reply by bluebill1049** [reply]
Nah all good matt. glad it's working out. have a good weekend.

**Reply by SusmitaSaha12** [reply]
Hi Bill,

I am literally stuck with integrating react-draft-wysiwyg with react-hook-form and not getting any way outs. Below is my code snippet:

import React from "react";
import { useForm, useController } from "react-hook-form";
import { EditorState } from "draft-js";
import { Editor } from "react-draft-wysiwyg";
import "react-draft-wysiwyg/dist/react-draft-wysiwyg.css";

export default function List() {
  const { handleSubmit, control } = useForm();
  const { field } = useController({
    name: "editorField",
    control,
    defaultValue: EditorState.createEmpty()
  });

  const onSubmit = handleSubmit((data) => {
    console.log("editor value"+data);
  });

  return (
    <form onSubmit={onSubmit}>
      <Editor
        editorState={field.value}
        toolbarClassName="toolbarClassName"
        wrapperClassName="wrapperClassName"
        editorClassName="editorClassName"
        onEditorStateChange={field.onChange}
        onBlur={field.onBlur}
        editorStyle={{ height: "200px" }}
      />

      <hr />
      <button type="submit">Submit</button>
    </form>
  );
}

When i run this code it gives **editor value[object Object]**. 
Earnestly request your kind help at the earliest as i am badly stuck with it and already the project delivery timeline is reached. Please help.

**Reply by khanhleemtp** [reply]
> https://codesandbox.io/s/wizardly-williams-2cjoe?file=/src/App.js
when i use code, i received warning: 
Can't call setState on a component that is not yet mounted. This is a no-op, but it might indicate a bug in your application. Instead, assign to `this.state` directly or define a `state = {};` class property with the desired state in the r component.
and not change 'content_editor'. Please help u.

**Reply by SusmitaSaha12** [reply]
hi khanhleemtp,
u can use this
https://codesandbox.io/s/editor-pv9ui

**Reply by khanhleemtp** [reply]
> hi khanhleemtp, u can use this https://codesandbox.io/s/editor-pv9ui

i think it's woking. tks u

**Reply by ngurahyudi** [reply]
Hi @bluebill1049 I'm failed to reset the react draft WYSIWYG value using reset method of RHF, any clue? Thanks

**Reply by bluebill1049** [reply]
@ngurahyudi would you like to send a separate GitHub question with a codesandbox? I can take a look at it.

**Reply by ngurahyudi** [reply]
> @ngurahyudi would you like to send a separate GitHub question with a codesandbox? I can take a look at it.

sure, im working on it.. thanks Bill

**Reply by ngurahyudi** [reply]
> Hi @bluebill1049 I'm failed to reset the react draft WYSIWYG value using reset method of RHF, any clue? Thanks
vvvvvvvvvv
answered [here](https://github.com/react-hook-form/react-hook-form/discussions/7814#discussion-3875167)

**Comment by SusmitaSaha12** [comment]
Also, i tried with the code below:

import React, { useState } from "react";

// Components
import { EditorState, convertToRaw } from "draft-js";
import { Editor } from "react-draft-wysiwyg";
//import draftToHtml from "draftjs-to-html";

import "react-draft-wysiwyg/dist/react-draft-wysiwyg.css";
//import "./WYSIWYG.scss";

const Summerform = props => {
  console.log("props"+props);

  const [editorState, setEditorState] = useState(EditorState.createEmpty());
  const onEditorStateChange = editorState => {
    setEditorState(editorState);
    console.log("props editor", editorState);
    // props.onChange(editorState);
    // return input.onChange(
    // 	draftToHtml(convertToRaw(editorState.getCurrentContent()))
    // );
  };

  return (
    <React.Fragment>
      <div className="editor">
        <Editor
          editorState={editorState}
          wrapperClassName="wrapper-class"
          editorClassName="editor-class"
          onEditorStateChange={onEditorStateChange}
        />
      </div>
    </React.Fragment>
  );
};

export default Summerform;



But here on running the code above it is giving undefined.
Console log:

props editor EditorState {_immutable: Record}_immutable: Record {_map: Map, __ownerID: undefined}[[Prototype]]: Object


Please please help.

**Comment by SusmitaSaha12** [comment]
Hi Matt, Bill

Please help. earnest request.

**Comment by SusmitaSaha12** [comment]
hi mattbaratella & bluebill1049,

Please help  earnest request. I am literally stuck with integrating react-draft-wysiwyg with react-hook-form and not getting any way outs.

**Reply by mattbara** [reply]
@SusmitaSaha12 I am sorry but I am in the middle of a big release at work and have no time to look into someone else code. Plus this is not my library and I haven't checked the new updates in ages. I think Bill is your best option. I would suggest who ever has an issue with it, to provide a code sandbox link with the full code and all the libraries. It will be easier to debug

**Reply by SusmitaSaha12** [reply]
Hi Matt @devolab ,

Sorry to bother you in the mid of your release but i am badly in need of your help. The link that you provided in your codesandbox  i.e. https://codesandbox.io/s/trusting-mountain-k24ys?file=/src/WYSIWYG.js, i have checked and the same is working OK in codesandbox environment. 
**But surprisingly the same code is not working when i create a separate project and a single file so that there is no conflict with any other part of project code.** The problem that i have found is the props when getting initialized is holding value undefined. and that is the main reason. Cant figure out what's the reason.

**Below is my code snippet and o/p. Just to re-iterate, my code snippet means, your code only given in your codesandbox.**

const WYSIWYGEditor = props => {
  console.log("props"+props);

  const [editorState, setEditorState] = useState(EditorState.createEmpty());
  const onEditorStateChange = editorState => {
    setEditorState(editorState);
    console.log("props editor", props);
    // props.onChange(editorState);
    // return input.onChange(
    // 	draftToHtml(convertToRaw(editorState.getCurrentContent()))
    // );
  };
O/P is 
props editor EditorState {_immutable: Record}_immutable: Record {_map: Map, __ownerID: undefined}
[[Prototype]]: Object
 props[object Object]

Our codesandbox link : https://codesandbox.io/s/reverent-shirley-pv9ui?file=/src/App.js
Same code has been copied and pasted in my codesandbox but the same threw error. **e.render** is not a function.
I have added 2 lines to fix the bug and then when i run the code, it gives undefined. No props is also getting initialized. 

But the props should be like value what i am seeing in your code sandbox:
PROPS ==>  
{name: "editor_content", onChange: ƒ handleChange(), onBlur: ƒ onBlur(), value: "<p></p>
"}


Cant figure out and earnestly seeking help.  How the same code can behave in a complete different manner in two different sandbox?

@bluebill1049 @mattbaratella @devolab 

Request you all to please help me and oblige so that i can finish my project delivery. Please. If i would have not been in extreme urgency and project delivery pressure, surely would have not disturbed. Please help.

**Reply by bluebill1049** [reply]
Is this working for you: https://codesandbox.io/s/recursing-zeh-sosjd?file=/src/WYSIWYG.js

**Reply by mattbara** [reply]
@SusmitaSaha12 you are looking in the wrong file. The file you need to look into is the one which calls the component < WYSIWYGEditor ......>. If you are saying that the props passed to < WYSIWYGEditor /> are undefined, why are you looking at this component? You need to go up 1 hierarchy and debug thep props you are passing BEFORE they get passed. Debug backwards. If you check the codesandbox, the file you are looking for is something similar to App.js. I hope it helps.

**Reply by mattbara** [reply]
@SusmitaSaha12 look @bluebill1049 codesandbox. The Code seems much better than the one I wrote. If that doesn't work I would suggest you to put everything on a github repo and provide a full installation of your app. But I won't be able to look at any for at least 2 weeks.

**Reply by SusmitaSaha12** [reply]
Thank You  bluebill1049 & devolab for helping . bluebill1049 your codesand box is working on my project .
THANK YOU SO MUCH

**Reply by bluebill1049** [reply]
no worries @SusmitaSaha12 thank you @devolab for helping 🙏 ❤️

**Comment by tannvdts** [comment]
@mattbaratella  You can use the library https://www.npmjs.com/package/@tonz/react-draft-wysiwyg-input
This library wapper eact-draft-wysiwyg with behaviour like input html:
```

import {RichTextInput} from '@tonz/react-draft-wysiwyg-input'

import '@tonz/react-draft-wysiwyg-input/style.css';

<RichTextInput disabled={disabled} {...register('desc')} />
```

**Reply by krismp** [reply]
I'm using this in my NextJS app, I'm using typescript, and seems it can't find any type definitions
